keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2868 @@
+    messy and complicated to reconstruct.
+- Traverse across fields safely? CEL offers the `all()` macro, so traversals
+  like `spec.initContainers.all(c, c.name)` are possible, but can be subtly
+  incorrect because of optional fields (`initContainers` in this case), which
+  require `has()` checks, e.g.: `!has(spec.initContainers) || spec.initContainers.all(c, c.name)`.
+
+CRD Validation Rules solved these problems by allowing validation rules to
+attached to any schema in the OpenAPIv3. The validation rules are scoped to
+whatever location in the OpenAPIv3 they are attached.
+
+We propose using a simple schema path format. The purpose of the path is to
+uniquely identify a schema from the root of a CRDs OpenAPIv3 schema.
+
+Example:
+
+```
+spec.initContainers         # Schema of initContainers array
+spec.initContainers[*]      # Schema of the items of the initContainers array
+spec.initContainers[*].name # Schema of the name of initContainers
+```
+
+For example, to validate all containers:
+
+```yaml
+  validations:
+    - scope: "spec.containers[*]"
+      expression: "scope.name.startsWith('xyz-')"
+      message: "{1} does not start with 'xyz'"
+      messageArgs: ["scope.name"]
+```
+
+To make it possible to access the path information in the scope, we can offer a
+way to bind varables to the map and list indices in the path, e.g.:
+
+```
+spec.x[xKey].y[yIndex].field
+```
+
+```yaml
+  validations:
+    - scope: "x[xKey].y[yIndex].field"
+      expression: "scope.startsWith('xyz-')"
+      message: "{1}, {2}: some problem"
+      messageArgs: ["scopePath.xKey", "scopePath.yIndex"]
+```
+
+Prior art:
+
+- cel-policy-template's offer a
+  [`range`](https://github.com/google/cel-policy-templates-go/blob/master/test/testdata/map_ranges/template.yaml)
+  feature that allows a CEL expression to be scoped to each entry of a map or
+  item of an array. Multiple ranges can be combined to traverse a complex
+  object.
+- Kyverno [foreach
+  declarations](https://kyverno.io/docs/writing-policies/validate/#foreach), use
+  [JMESPath](https://jmespath.org/) to query for the elements that are then
+  validated.
+
+Note: We considered extending to a list of scopes, e.g.:
+
+```yaml
+  validations:
+    - scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
+      expression: "scope.name.startsWith('xyz-')"
+      message: "{1} does not start with 'xyz'"
+      messageArgs: ["scope.name"]
+```
+
+But feedback was this is signficantly more difficult to understand.
+
+### Message formatting alternatives
+
+Alternative: offer a CEL expression
+
+```yaml
+- expression: "..."
+  messageExpression: "string(object.int1) + ' is less than ' + string(object.int2)"
+```
+
+Cons:
+
+- CEL requires explicit casts to string
+- Plain string message support is a bit messy. Options:
+  - Use CEL always: `"'this is a simple message'"`
+  - Offer both plain string (`message`) and CEL (`messageExpression`)
```

Comment:
I actually like this option of two fields. And, I think the introduction of a `string.format()` function in CEL would address the first con of requiring explicit casts.

Reply:
If we can get `string.format()` in by beta I think I'm comfortable with this approach. We can add `messageExpression` to CRD validation as well.